ibis/expr/api.py

```diff
@@ -4193,27 +4193,54 @@ def mutate(table, exprs=None, **mutations):
         for name, expr in sorted(mutations.items(), key=operator.itemgetter(0))
     )
 
-    for expr in exprs:
-        if expr.get_name() and isinstance(expr, ir.DestructColumn):
-            raise com.ExpressionError(
-                f"Cannot name a destruct column: {expr.get_name()}"
-            )
-
-    by_name = collections.OrderedDict(
-        (expr.get_name(), expr) for expr in exprs
-    )
+    selection_exprs_by_name = _get_selection_exprs_by_name(exprs)
     columns = table.columns
-    used = by_name.keys() & columns
+    used = selection_exprs_by_name.keys() & columns
 
     if used:
         proj_exprs = [
-            by_name.get(column, table[column]) for column in columns
-        ] + [expr for name, expr in by_name.items() if name not in used]
+            selection_exprs_by_name.get(column, table[column])
+            for column in columns
+        ]
+        # We do an additional check below that the expr is not already in
+        # proj_exprs because multiple columns can be encapsulated within a
+        # DestructColumn, and without this check, we could be duplicating the
+        # same DestructColumn multiple times in proj_exprs.
+        # Example: DestructColumn(['existing_column', 'new_column'])
+        # The DestructColumn will be added to proj_exprs above because
+        # 'existing_column' is being overwritten, and therefore 'new_column'
+        # is also already represented in proj_exprs, so we shouldn't add the
+        # DestructColumn expr below even though 'new_column' isn't in 'used'.
+        [
```

Comment:
Thinking out lound here:

Let say original tables have

"colA" "existingColumn" "ColB"

and your destruct column is:

DestructColumn(['existing_column', 'new_column'])

The selection_exprs_by_name would be
["existing_column", "new_column"]

And used is
["existing_column"]

In this case, the logic can be:

Split the exprs into (a) a dict (ovewritten_columns) from the the first column that an expr overrides to the expr (b) list of exprs that don't override anything. and the final proj_exprs would be

[table[column] if column not in ovewritten_columns else ovewritten_columns[column] for column in columns] + [list of exprs that don't override]

Reply:
In your example, the DestructColumn(['existing_column', 'new_column']) would have an entry
{'existing_column': DestructColumn(['existing_column', 'new_column'])} and the second list would be empty

Reply:
And when you iterate through table.columns, you would just replace the ones that are ovewritten.

Reply:
Got it, yep that definitely makes the logic simpler rather than iterating back over the exprs.